kyuubi-common/src/main/scala/org/apache/kyuubi/operation/log/LogDivertAppender.scala

```diff
@@ -21,43 +21,238 @@ import java.io.CharArrayWriter
 
 import scala.collection.JavaConverters._
 
-import org.apache.log4j._
-import org.apache.log4j.spi.{Filter, LoggingEvent}
+import org.apache.logging.log4j.{Level, LogManager, Marker}
+import org.apache.logging.log4j.core.{Filter, LogEvent, Logger, StringLayout}
+import org.apache.logging.log4j.core.Filter.Result
+import org.apache.logging.log4j.core.LifeCycle.State
+import org.apache.logging.log4j.core.appender.{AbstractWriterAppender, ConsoleAppender, WriterManager}
+import org.apache.logging.log4j.core.config.Property
+import org.apache.logging.log4j.core.layout.PatternLayout
+import org.apache.logging.log4j.message.Message
 
-class LogDivertAppender extends WriterAppender {
+class LogDivertAppender(
+    name: String,
+    layout: StringLayout,
+    filter: Filter,
+    ignoreExceptions: Boolean,
+    immediateFlush: Boolean,
+    writer: CharArrayWriter)
+  extends AbstractWriterAppender[WriterManager](
+    name,
+    layout,
+    filter,
+    ignoreExceptions,
+    immediateFlush,
+    Property.EMPTY_ARRAY,
+    new WriterManager(writer, name, layout, true)) {
+  def this() = this(
+    "KyuubiEngineLogDivertAppender",
+    LogDivertAppender.initLayout(),
+    null,
+    false,
+    true,
+    LogDivertAppender.writer)
 
-  final private val writer = new CharArrayWriter
-
-  final private val lo = Logger.getRootLogger
-    .getAllAppenders.asScala
-    .find(_.isInstanceOf[ConsoleAppender])
-    .map(_.asInstanceOf[Appender].getLayout)
-    .getOrElse(new PatternLayout("%d{yy/MM/dd HH:mm:ss} %p %c{2}: %m%n"))
-
-  setName("KyuubiSparkSQLEngineLogDivertAppender")
-  setWriter(writer)
-  setLayout(lo)
-
-  addFilter { _: LoggingEvent =>
-    if (OperationLog.getCurrentOperationLog == null) Filter.DENY else Filter.NEUTRAL
-  }
+  addFilter(new NameFilter())
 
   /**
-   * Overrides WriterAppender.subAppend(), which does the real logging. No need
+   * Overrides AbstractWriterAppender.append(), which does the real logging. No need
    * to worry about concurrency since log4j calls this synchronously.
    */
-  override protected def subAppend(event: LoggingEvent): Unit = {
-    super.subAppend(event)
+  override def append(event: LogEvent): Unit = {
+    super.append(event)
     // That should've gone into our writer. Notify the LogContext.
     val logOutput = writer.toString
     writer.reset()
     val log = OperationLog.getCurrentOperationLog
     if (log != null) log.write(logOutput)
   }
+
+  class NameFilter extends Filter {
```

Comment:
extends NeutralFilter?

Reply:
do you mean 
```
class NeutralFilter extends Filter?
```

There is no class named `NeutralFilter`.

Reply:
sorry my mistake, using `org.apache.logging.log4j.core.filter.AbstractFilter`